src/plugins/discourse/fetch.js

```diff
@@ -149,23 +149,119 @@ export class Fetcher implements Discourse {
     return this._fetchImplementation(fullUrl, fetchOptions);
   }
 
-  async latestTopicId(): Promise<TopicId> {
-    const response = await this._fetch("/latest.json?order=created");
+  async categoryDefinitionTopicIds(): Promise<TopicId[]> {
+    const topicIdRE = new RegExp("/t/[\\w-]+/(\\d+)$");
+    const urls = [];
+    const withSubs = [];
+
+    // Root categories
+    const response = await this._fetch(
+      `/categories.json?show_subcategory_list=true`
+    );
     failIfMissing(response);
     failForNotOk(response);
-    const json = await response.json();
-    if (json.topic_list.topics.length === 0) {
-      throw new Error(`no topics! got ${stringify(json)} as latest topics.`);
+    const {categories: rootCategories} = (await response.json()).category_list;
+    for (const cat of rootCategories) {
+      if (cat.topic_url != null) {
+        urls.push(cat.topic_url);
+      }
+      if (cat.subcategory_ids) {
+        withSubs.push(cat.id);
+      }
+    }
+
+    // Subcategories
+    for (const rootCatId of withSubs) {
+      const subResponse = await this._fetch(
+        `/categories.json?show_subcategory_list=true&parent_category_id=${rootCatId}`
+      );
+      failIfMissing(subResponse);
+      failForNotOk(subResponse);
+      const {
+        categories: subCategories,
+      } = (await subResponse.json()).category_list;
+      for (const cat of subCategories) {
+        if (cat.topic_url != null) {
+          urls.push(cat.topic_url);
+        }
+      }
     }
-    return json.topic_list.topics[0].id;
+
+    const ids = urls.map((url) => {
+      const match = topicIdRE.exec(url);
+      if (match == null) {
+        throw new Error(
+          `Encountered topic URL we failed to parse it's TopicId from: ${url}`
+        );
+      }
+      return Number(match[1]);
+    });
+    const uniqueIds = Array.from(new Set(ids).values());
+    uniqueIds.sort((a, b) => a - b);
+    return uniqueIds;
   }
 
-  async latestPosts(): Promise<Post[]> {
-    const response = await this._fetch("/posts.json");
-    failIfMissing(response);
-    failForNotOk(response);
-    const json = await response.json();
-    return json.latest_posts.map(parsePost);
+  async topicsBumpedSince(sinceMs: number): Promise<Topic[]> {
+    const topics = [];
+    let pastSince = false;
+    let morePages = true;
+    let page = 0;
+
+    while (!pastSince && morePages) {
```

Comment:
idea: instead of `while (!pastSince && morePages)` how about `while (lastUnpinnedTimestamp <= sinceMs && morePages)` (with the obvious semantics for `lastUnpinnedTimestamp`).

As it currently is, the logic for the control loop is split between the while declaration, and the bottom of the function where `pastSince` is evaluated. With this change, the conceptual control logic is declared all in one place, with only the implementation of tracking `lastUnpinnedTimestamp` at the bottom.

Reply:
I'm generally a bit miffed with this code scope, it feels like there's a lot going on, including a lot of mutation of state outside the scope, and it's hard for me to feel confident that i understand it. 

what if we define helper functions like:

`processTopic`: validates an individual topic and returns the parsed Topic, along with any extra summary info
`addPage`: collects all the topics a whole response page, also returns the information needed to decide whether there are more pages 

then the outer block can just call `addPage` in a loop, and it will be easier for the reader to focus on individual pieces of functionality